updates on lms doc (#110)
TiarkRompf/lms-clean: doc/main/scala/lms/core/backend.md

The `lms/core/backend.scala` file contains several important aspects of the LMS IR. In `object Backend`, we define the basic building blocks of the IR and effect-related data structures. In `class GraphBuilder`, `class GraphBuilderOpt` and `case class Graph`, we introduce how new nodes are added into a graph IR and how effects are handled and used to calculate dependencies. We also define `Phase` as a framework for graph-to-graph transformation and introduce a dead code elimination (DCE) phase `DeadCodeElimCG`.

# Backend
The core LMS IR is defined in `object Backend`. We first introduce the basic building blocks of the IR:

``` scala
abstract class Def // Definition: used in right-hand-side of all nodes
abstract class Exp extends Def
case class Sym(n: Int) extends Exp   // Symbol
case class Const(x: Any) extends Exp // Constant
case class Block(in: List[Sym], res: Exp, ein: Sym, eff: EffectSummary) extends Def

case class Node(n: Sym, op: String, rhs: List[Def], eff: EffectSummary)
```

> `Def` represents a definition, which is used in the RHS of all nodes. A definition can be an `Exp` or a `Block`. `Exp` can be a `Sym` or a `Const`. `Sym` represents symbols in the IR identified by a unique integer. For example, `Sym(2)` represents `x2` in the generated code. `Const` represents immediate values, such as integers (e.g. `Const(0)`), strings (e.g. `Const(hello)`), etc.

Reviewer: ```suggestion
`Def` represents a definition, which is used in the RHS of all nodes. A definition can be an `Exp` or a `Block`. `Exp` can be a `Sym` or a `Const`. `Sym` represents symbols in the IR identified by a unique integer. For example, `Sym(2)` represents `x2` in the generated code. `Const` represents immediate values, such as integers (e.g. `Const(0)`), strings (e.g. `Const("hello")`), etc.
```

> `Node` represents statements in the IR. The first field `n` of a node is a `Sym` corresponding to the LHS of a statement. The second field `op` is the operator of the statement. We use strings as operators since they permit easy extension of various kinds of nodes such as `-`, `print`, etc. The third field is the list of operands (RHS) of a node. Let's ignore the last field `EffectSummary` for now, which will be covered later. For example, the statement `val x3 = x2 + 1` would become

Reviewer: ```suggestion
`Node` represents statements in the IR. The first field `n` of a node is a `Sym` corresponding to the identity of a statement. The second field `op` is the operator of the statement. We use strings as operators since they permit easy extension of various kinds of nodes such as `-`, `print`, etc. The third field is the list of operands (RHS) of a node. Let's ignore the last field `EffectSummary` for now, which will be covered later. For example, the statement `val x3 = x2 + 1` would become
```

}
```

We can achieve the same optimization using an expensive dataflow analysis on a tree-based IR. But you will see in later sections that our approach is more lightweight and easier to implement.

### Data and Control Dependencies

From the previous example, hopefully the reader has gained a basic understanding of why we use sea-of-nodes IR. We briefly mentioned that the nodes are scheduled into blocks using certain criteria. What are these criteria? 

The criteria we use in LMS IR are *dependencies*. Simply speaking, dependencies capture information flow in the source program that we must respect. Given the order of nodes in the source program defined by the programmer, we can see that certain nodes can be reordered or moved in/out of scopes without changing the semantics of the program, but certainly, that's not true for all nodes. Dependency information tells us the order of nodes we must keep when we schedule nodes. There are two kinds of dependencies that we care about in LMS IR:
1. the data dependency and
2. the control dependencies.

The data dependency is straightforward. If a node uses an `Exp`, then the node

> depends on that `Exp` and the node that generates that `Exp`. Data dependency is captured explicitly by our definition of `Node`. If the `rhs` of node A contains node B, then A is data-dependent on B. For instance, in the following code snippet:

Reviewer: ```suggestion
depends on that `Exp` and the node that generates that `Exp`. Data dependency is captured explicitly by our definition of `Node`. If the `rhs` of node A refers to node B, then A is data-dependent on B. For instance, in the following code snippet:
```

print(x)
```

In this example, the node `x = 10` is not *data-depended* by the print statement.
However, it is *control-depended* because the value of `x` is mutated by it. However,
the `x = 3` node can be removed without breaking the code semantics. The order of
graph construction (the fact that the user did `x = 3` before `x = 10`) is the base of our control dependency analysis.

### Effects and EffectSummary

How can we obtain data and control dependency information from the source program? As we mentioned before, data dependency is already captured by the definition of `Node`. However, it is not clear how to collect control dependencies. It turns out that we can track the *effects* of nodes and blocks and then compute control dependencies from these effects. To make our analysis more fine-grained (so that more interesting code reordering and optimization can be introduced), the first step is to realize that there are different kinds of effects. The effect of printing is clearly not so related to the effect of variable reading. Thus there should be no scheduling enforcement between them.

> In LMS IR, we track *read* and *write* effects. They are categorized in the following cases:

Reviewer: I guess it would be good to point out that these "phantom" effect keys (such as STORE/CTRL) can be user-defined too.

on the block input `x`. In this case, we let the print node depends on the `ein` of the
block, so that it will be scheduled in the block, but not before.

### From Effects to Dependencies

After collecting effects, we need to compute dependencies from them. The
rules for computing dependencies from effects are listed below:

1. Read-After-Write (RAW): there should be a hard dependency from R to W,
   since the correctness of the R depends on the W to be scheduled)
2. Write-After-Read (WAR): there should be a soft dependency from W to R,
  since the correctness of the W does not depend on the R to be scheduled, but the order of them matters.
3. Write-After-Write (WAW): the first idea is to generate a soft dependency,
  since the second W simply overwrites the first W.

>   However, write to array is more complicated, such as arr(0) = 1; arr(1) = 2,

Reviewer: ```suggestion
  However, write to array is more complicated, such as `arr(0) = 1; arr(1) = 2`,
```

  }
}
```
We first add the top-most block node of the graph to `reach`. If the block returns a `Sym`, we add that symbol to both `live` and `used`. We then add the effect and data input symbols to `used`. Then, we traverse all nodes in the input graph in reverse order. If a node is reachable, we add the `valueSyms` of that node to `live`, and `hardSyms` to `reach`. The `valueSym` collects all used values of a node, including the symbols and block inputs of the RHS. The `hardSym` collects hard dependencies of a node. Lastly, we add the current reachable node to `newNodes`.

### Remove Unused Variables
```scala
for (d <- newNodes.reverseIterator) {
  if (used(d.n)) {
    used ++= valueSyms(d)
  } else if (d.eff.hasSimpleEffect || d.eff.wkeys.exists(used)) {
    used += d.n
    used ++= valueSyms(d)
  }
}
```
After getting all reachable nodes in `newNodes`, we continue collect used variables by traversing `newNodes` in reverse order. For each node `d`, if the symbol of `d` is in `used`, we add the `valueSym` of `d` into `used` as well. If `d` is not in `used` but has simple effects (e.g. print) or writes to a used symbol, we also add the symbol and `valueSym` of `d` into `used`.

> ### Recreate the Graph

Reviewer: ```suggestion
### Reconstruct the Graph 
```

```scala
var newGlobalDefsCache = Map[Sym,Node]()
newNodes = for (d <- newNodes if used(d.n)) yield {
  newGlobalDefsCache += d.n -> d
  if (d.op == "staticData") statics += d
  if (fixDeps)
    d.copy(rhs = d.rhs.map {
      case b: Block => b.copy(eff = b.eff.filter(used))
      case o => o
    }, eff = d.eff.filter(used))
  else
    d
}
val newBlock = if (fixDeps)
  g.block.copy(eff = g.block.eff.filter(used))
else
  g.block

Graph(newNodes, newBlock, newGlobalDefsCache)
```

> Lastly, we recreate the graph based on the information we collected earlier. We only consider nodes in `newNodes` that are also in `used`. For each such node, we update `newGlobalDefsCache` to map the symbol of the node to the node itself. We then remove unused effect symbols from the RHS and `EffectSummary` of the node. We do the same thing to the top-most block and return a new `Graph` with the updated information.

Reviewer: ```suggestion
Lastly, we reconstruct the graph based on the information we collected earlier. We only consider nodes in `newNodes` that are also in `used`. For each such node, we update `newGlobalDefsCache` to map the symbol of the node to the node itself. We then remove unused effect symbols from the RHS and `EffectSummary` of the node. We do the same thing to the top-most block and return a new `Graph` with the updated information.
```
